lib/pacemaker/pcmk_result_code.c

```diff
@@ -0,0 +1,215 @@
+ *                   \c pcmk_result_crm_exitcode.
+ * \param[out] name  Where to store the result code's name
+ * \param[out] desc  Where to store the result code's description
+ *
+ * \return Standard Pacemaker return code
+ *
+ * \note This function should return \c pcmk_rc_ok even if the result code is
+ *       unknown. It will return an error only if the lookup function returns
+ *       \c NULL, which should never happen.
+ */
+static int
+get_strings(int code, pcmk_result_type_t type, const char **name,
+            const char **desc)
+{
+    CRM_ASSERT((type == pcmk_result_legacy) || (type == pcmk_result_pcmk_rc)
+               || (type == pcmk_result_crm_exitcode));
+
+    CRM_ASSERT((name != NULL) && (desc != NULL));
+
+    switch (type) {
+        case pcmk_result_legacy:
+            *name = pcmk_errorname(code);
+            *desc = pcmk_strerror(code);
+            break;
+        case pcmk_result_pcmk_rc:
+            *name = pcmk_rc_name(code);
+            *desc = pcmk_rc_str(code);
+            break;
+        case pcmk_result_crm_exitcode:
+            *name = crm_exit_name((crm_exit_t) code);
+            *desc = crm_exit_str((crm_exit_t) code);
+            break;
+        default:    // Make the compiler happy
+            *name = NULL;
+            *desc = NULL;
+            break;
+    }
+
+    if (*name == NULL) {
+        return ENXIO;
+    }
+    return pcmk_rc_ok;
+}
+
+/*!
+ * \internal
+ * \brief Display the name and/or description of a result code
+ *
+ * \param[in,out] out        Output object
+ * \param[in]     code       The result code
+ * \param[in]     type       Interpret \c code as this type of result code.
+ *                           Supported values: \c pcmk_result_legacy,
+ *                           \c pcmk_result_pcmk_rc,
+ *                           \c pcmk_result_crm_exitcode.
+ * \param[in]     show_name  Whether to show the result code's name
+ * \param[in]     show_desc  Whether to show the result code's description
+ *
+ * \return Standard Pacemaker return code
+ *
+ * \note At least one of \c show_name and \c show_desc must be set to \c true.
+ */
+int
+pcmk__show_result_code(pcmk__output_t *out, int code, pcmk_result_type_t type,
+                       bool show_name, bool show_desc)
+{
+    int rc = pcmk_rc_ok;
+    bool quiet_orig = out->is_quiet(out);
+
+    const char *name = NULL;
+    const char *desc = NULL;
+
+    // Otherwise we're displaying nothing
+    CRM_ASSERT(show_name || show_desc);
+
+    rc = get_strings(code, type, &name, &desc);
+    if (rc != pcmk_rc_ok) {
+        out->err(out, "Error looking up result code %d", code);
+        return rc;
+    }
+
+    /* Enable quiet temporarily to suppress the code in the output (if quiet is
+     * supported)
+     */
+    out->quiet = true;
+    out->message(out, "result-code", code, (show_name? name : NULL),
+                 (show_desc? desc : NULL));
+    out->quiet = quiet_orig;
+
+    return rc;
```

Comment:
I'm on the fence about whether we should return an error code like `ENXIO` (mapped to `CRM_EX_NOSUCH`) if we our lookup returns an unknown error. Currently, we're returning `pcmk_rc_ok` because the query didn't do anything off-the-wall like returning a null pointer -- it printed the string "Unknown" or something like that. Further, `crm_error` doesn't return an error in this case.

But it seems like an arbitrary caller might want to check for an error code to know whether we found a real value for the code we're querying, or whether we simply printed "Unknown".

`crm_error` can always ignore an error code from `pcmk__show_return_code()`. However, `crm_error` might want to do something in a hypothetical case where `get_strings()` gives us a null pointer, even if it ignores the result "Unknown". Could use two different error codes but `ENXIO` seems like the closest match for both. May have to make a new one.

Reply:
`crm_error` itself probably should return an error in case of "Unknown", but changing the exit codes could be a regression.